### DISCUSSION TITLE
Confused about how to work with APIs

### DISCUSSION THREAD

**Discussion by Shubhamn96** [microsoft/AirSim#3241] [Support Q&A] [answered]
So I am working with AirSim for my capstone project and for one part, I want to be able to read the sensor data. In the documentation, I see that I can use getGpsdata() but my question is where should I write the code for collecting the information? Do I write this stuff in settings.json or should I create a new script altogether? I am very new to this side of programming and don't know how to work with APIs that well

**Comment by ahmed-elsaharti** [comment, marked as answer]
Hello @Shubhamn96 , the [APIs](https://github.com/microsoft/AirSim/blob/master/docs/apis.md) could be used in a [C++](https://github.com/microsoft/AirSim/blob/master/docs/apis_cpp.md) project or a [Python](https://github.com/microsoft/AirSim/blob/master/docs/apis.md#python-quickstart) script.

If you're new to programming I suggest that you start by using the Python APIs. A good starting point is the demo python scripts under [PythonClient/Multirotor](https://github.com/microsoft/AirSim/tree/master/PythonClient/multirotor). You might want to have a look at the code they contain. To get started you will need to [install Python](https://www.youtube.com/watch?v=UvcQlPZ8ecA), [install PIP](https://www.youtube.com/watch?v=Ko9b_vC6XY0), [install the airsim python library](https://github.com/microsoft/AirSim/blob/master/docs/apis.md#installing-airsim-package) then follow [these steps ](https://github.com/microsoft/AirSim/blob/master/docs/apis.md#python-quickstart) to run hello_drone.py to test its functionality out.

Hope this helps

**Reply by Shubhamn96** [reply]
So am I creating a new .py file, importing all the necessary libraries, and making my own code?

**Reply by ahmed-elsaharti** [reply]
Precisely. all you will need to import for the basic functionality would be the `airsim` library through `import airsim`.
Then your code would be able to connect to the drone (I'm assuming you're going to use the drone simulation here) through
```python
client = airsim.MultirotorClient()
client.confirmConnection()
```
enable API control through
```python
client.enableApiControl(True)
```
then you can arm/disarm, takeoff etc, etc
```python
client.armDisarm(True)
client.takeoffAsync().join()
```
__Note__ : adding `.join()` to a motion API call makes the python code wait till the command is executed before proceeding to the next line

**Reply by Shubhamn96** [reply]
And should the new file be under the AirSim directory or in my environment directory

**Reply by ahmed-elsaharti** [reply]
@Shubhamn96 That would not matter if you've installed the airsim library.
If you didn't, I would keep it under `AirSim\PythonClient\multirotor` and would import `setup_path` before importing `airsim`

**Reply by Shubhamn96** [reply]
> **Note** : adding `.join()` to a motion API call makes the python code wait till the command is executed before proceeding to the next line

I want to use an RC controller to control the simulation, so could I initiate a while loop that keeps running and storing sensor information either while:
A. The RC controller is connected, once it is connected the simulation stops collecting data 
B. The simulation itself is running, one ESC is hit, the simulation and data collection ends

**Reply by ahmed-elsaharti** [reply]
Not entirely sure I understand what you're trying to do here. But in general, if you dont execute `client.enableApiControl(True)` or if you set it to False you'll have RC control (assuming you're using a USB controller here of some sort) and you can have your loop collect the data normally without interrupting RC movement.
Also, in general, when you hit ESC the simulation would end and if you try to grab data through the API within the loop It'll probably either crash or raise an exception. The safest way imo to go about this is to have the loop exit on keypress or something then stop the simulation.
